lib/ui/channel_buffers.dart

```diff
@@ -0,0 +1,173 @@
+      if (_dropItemCallback != null) {
+        _dropItemCallback(item);
+      }
+      overflow = true;
+    }
+    _queue.addLast(val);
+    return overflow;
+  }
+
+  /// Returns null when empty.
+  T pop() {
+    return _queue.isEmpty ? null : _queue.removeFirst();
+  }
+
+  /// Returns the number of discarded items resulting from resize.
+  int resize(int newSize) {
+    int result = 0;
+
+    while (length > newSize) {
+      result += 1;
+      final T item = _queue.removeFirst();
+      if (_dropItemCallback != null) {
+        _dropItemCallback(item);
+      }
+    }
+
+    _capacity = newSize;
+
+    return result;
+  }
+}
+
+typedef DrainChannelCallback = Future<void> Function(ByteData, PlatformMessageResponseCallback);
+
+/// Storage of channel messages until the channels are completely routed,
+/// i.e. when a message handler is attached to the channel on the framework side.
+///
+/// Each channel has a finite buffer capacity and in a FIFO manner messages will
+/// be deleted if the capacity is exceeded.  The intention is that these buffers
+/// will be drained once a callback is setup on the BinaryMessenger in the
+/// Flutter framework.
+class ChannelBuffers {
+  /// By default we store one message per channel.  There are tradeoffs associated
+  /// with any size.  The correct size should be chosen for the semantics of your
+  /// channel.
+  ///
+  /// Size 0 implies you want to ignore any message that gets sent before the engine
+  /// is ready (keeping in mind there is no way to know when the engine is ready).
+  ///
+  /// Size 1 implies that you only care about the most recent value.
+  ///
+  /// Size >1 means you want to process every single message and want to chose a
+  /// buffer size that will avoid any overflows.
+  static const int kDefaultBufferSize = 1;
+
+  final Map<String, _RingBuffer<_StoredMessage>> _messages =
+    <String, _RingBuffer<_StoredMessage>>{};
+
+  _RingBuffer<_StoredMessage> _makeRingBuffer(int size) {
+    final _RingBuffer<_StoredMessage> result = _RingBuffer<_StoredMessage>(size);
+    result.dropItemCallback = _onDropItem;
+    return result;
+  }
+
+  void _onDropItem(_StoredMessage message) {
+    message.callback(null);
+  }
+
+  /// Returns true on overflow.
+  bool push(String channel, ByteData data, PlatformMessageResponseCallback callback) {
+    _RingBuffer<_StoredMessage> queue = _messages[channel];
+    if (queue == null) {
+      queue = _makeRingBuffer(kDefaultBufferSize);
+      _messages[channel] = queue;
+    }
+    final bool result = queue.push(_StoredMessage(data, callback));
+    if (result) {
+      // TODO(aaclarke): Update this message to include instructions on how to resize
+      // the buffer once that is available to users.
+      _Logger._printString('Overflow on channel: $channel.  '
+                           'Messages on this channel are being sent faster '
+                           'than they are being processed which is resulting '
```

Comment:
This wording provides a bit of a red herring. It'll make the user think that somehow the Dart code isn't keeping up whereas in reality, the Flutter code isn't running (either because their main entrypoint didn't call runApp or the main entrypoint never ran).